plume-common/src/activity_pub/inbox.rs

```diff
@@ -1,48 +1,560 @@
+        _ => None,
+    }
+}
+
+/// A trait for ActivityPub objects that can be retrieved or constructed from ID.
+///
+/// The two functions to implement are `from_activity` to create (and save) a new object
+/// of this type from its AP representation, and `from_db` to try to find it in the database
+/// using its ID.
+///
+/// When dealing with the "object" field of incoming activities, `Inbox` will try to see if it is
+/// a full object, and if so, save it with `from_activity`. If it is only an ID, it will try to find
+/// it in the database with `from_db`, and otherwise dereference (fetch) the full object and parse it
+/// with `from_activity`.
+pub trait FromId<C>: Sized {
+
+    /// The type representing a failure
+    type Error: From<InboxError<Self::Error>> + Debug;
+
+    /// The ActivityPub object type representing Self
+    type Object: activitypub::Object;
 
-    fn delete(&self, conn: &C) -> Result<A, Self::Error>;
-    fn delete_id(id: &str, actor_id: &str, conn: &C) -> Result<A, Self::Error>;
+    /// Tries to get an instance of `Self` from an ActivityPub ID.
+    ///
+    /// # Parameters
+    ///
+    /// - `ctx`: a context to get this instance (= a database in which to search)
+    /// - `id`: the ActivityPub ID of the object to find
+    /// - `object`: optional object that will be used if the object was not found in the database
+    ///   If absent, the ID will be dereferenced.
+    fn from_id(ctx: &C, id: &str, object: Option<Self::Object>) -> Result<Self, Self::Error> {
+        dbg!(object.is_some());
+        match Self::from_db(ctx, id) {
+            Ok(x) => Ok(x),
+            _ => match object {
+                Some(o) => Self::from_activity(ctx, o),
+                None => Self::from_activity(ctx, Self::deref(id)?),
+            },
+        }
+    }
+
+    /// Dereferences an ID
+    fn deref(id: &str) -> Result<Self::Object, Self::Error> {
+        dbg!(id);
+        reqwest::get(id)
+            .map_err(|_| InboxError::DerefError)
+            .and_then(|mut r| r.json().map_err(|_| InboxError::InvalidObject(None)))
+            .map_err(Into::into)
+    }
+
+    /// Builds a `Self` from its ActivityPub representation
+    fn from_activity(ctx: &C, activity: Self::Object) -> Result<Self, Self::Error>;
+
+    /// Tries to find a `Self` with a given ID (`id`), using `ctx` (a database)
+    fn from_db(ctx: &C, id: &str) -> Result<Self, Self::Error>;
 }
 
-pub trait WithInbox {
+/// Should be implemented by anything representing an ActivityPub actor.
+///
+/// # Type arguments
+///
+/// - `C`: the context to be passed to this activity handler from the `Inbox` (usually a database connection)
+pub trait AsActor<C>: Sized {
+    /// Return the URL of this actor's inbox
     fn get_inbox_url(&self) -> String;
 
-    fn get_shared_inbox_url(&self) -> Option<String>;
+    /// If this actor has shared inbox, its URL should be returned by this function
+    fn get_shared_inbox_url(&self) -> Option<String> {
+        None
```

Comment:
is this missing an implementation, or is this it? 😜

Reply:
No it is the default implementation: unless the implementation specifies a shared inbox URL, it is `None` (but in reality I think all our implementations override this behavior)